collector/fscache_linux.go

```diff
@@ -0,0 +1,264 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !nofscache
+// +build !nofscache
+
+package collector
+
+import (
+	"fmt"
+	"log/slog"
+	"os"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs"
+)
+
+const (
+	fscacheSubsystem = "fscache"
+)
+
+type fscacheCollector struct {
+	// Metrics
+	objectsAllocated *prometheus.Desc
+	objectsAvailable *prometheus.Desc
+
+	acquireAttempts *prometheus.Desc
+	acquireSuccess  *prometheus.Desc
+
+	lookupsTotal    *prometheus.Desc
+	lookupsPositive *prometheus.Desc
+	lookupsNegative *prometheus.Desc
+
+	invalidationsTotal *prometheus.Desc
+
+	updatesTotal *prometheus.Desc
+
+	relinquishesTotal *prometheus.Desc
+
+	attributeChangesTotal   *prometheus.Desc
+	attributeChangesSuccess *prometheus.Desc
+
+	allocationsTotal   *prometheus.Desc
+	allocationsSuccess *prometheus.Desc
+
+	retrievalsTotal    *prometheus.Desc
+	retrievalsSuccess  *prometheus.Desc
+	retrievalsNoBuffer *prometheus.Desc
+
+	storesTotal   *prometheus.Desc
+	storesSuccess *prometheus.Desc
+
+	cacheEventRetired *prometheus.Desc
+	cacheEventCulled  *prometheus.Desc
+
+	fs     procfs.FS
+	logger *slog.Logger
+}
+
+var _ prometheus.Collector = (*fscacheCollector)(nil)
+
+func init() {
+	registerCollector("fscache", defaultEnabled, func(logger *slog.Logger) (Collector, error) {
+		return NewFscacheCollector(logger)
+	})
+}
+
+// NewFscacheCollector returns a new Collector exposing fscache stats.
+func NewFscacheCollector(logger *slog.Logger) (*fscacheCollector, error) {
+	fs, err := procfs.NewFS(*procPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open procfs: %w", err)
+	}
+
+	return &fscacheCollector{
+		objectsAllocated: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, fscacheSubsystem, "objects_allocated_total"),
+			"Number of index cookies allocated (Cookies: idx=allocated/available/unused).",
+			nil, nil,
+		),
```

Comment:
We're using a pattern of package level Desc vars, rather than having `NewDesc()` in the `NewFooCollector()` function. This is to avoid allocating new Desc data when using the dynamic collector module mode.